plugins/storage/server/api/files.ts

```diff
@@ -0,0 +1,99 @@
+import Router from "koa-router";
+import { bytesToHumanReadable } from "@shared/utils/files";
+import env from "@server/env";
+import {
+  AuthenticationError,
+  InvalidRequestError,
+  NotFoundError,
+} from "@server/errors";
+import auth from "@server/middlewares/authentication";
+import validate from "@server/middlewares/validate";
+import { Attachment } from "@server/models";
+import { authorize } from "@server/policies";
+import { APIContext } from "@server/types";
+import { getFileFromRequest } from "@server/utils/koa";
+import { createRootDirForLocalStorage } from "../utils";
+import * as T from "./schema";
+
+createRootDirForLocalStorage();
+
+const router = new Router();
+
+router.post(
+  "files.create",
+  auth(),
+  validate(T.FilesCreateSchema),
+  async (ctx: APIContext<T.FilesCreateReq>) => {
+    if (!ctx.is("multipart/form-data")) {
+      throw InvalidRequestError("Request type must be multipart/form-data");
+    }
+    const file = getFileFromRequest(ctx.request);
+    if (!file) {
+      throw InvalidRequestError("Request must include a file parameter");
+    }
+
+    if (
+      env.FILE_STORAGE_UPLOAD_MAX_SIZE &&
+      file.size > env.FILE_STORAGE_UPLOAD_MAX_SIZE
+    ) {
+      throw InvalidRequestError(
+        `The selected file was larger than the ${bytesToHumanReadable(
+          env.FILE_STORAGE_UPLOAD_MAX_SIZE
+        )} maximum size`
+      );
+    }
```

Comment:
Time to extract this into a separate middleware `multipart({ maxAllowedFileSize: number })`? The file can be received as `ctx.input.file`.

Reply:
Great idea

Reply:
Created a separate issue [here](https://github.com/outline/outline/issues/5800).